### PR TITLE
refactor(compiler-cli): expose ngtools api separately

### PR DESCRIPTION
**Pull request by filipesilva**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
Private APIs needed for `@ngtools/webpack` were in exported from public APIs in `compiler-cli`.

## What is the new behavior?
Private APIs are exported exclusively through `@angular/compiler-cli/ngtools2`.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information
/cc @tbosch

### CODE REVIEW

File: packages/compiler-cli/ngtools.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+export {NgTools_InternalApi_NG_2 as __NGTOOLS_PRIVATE_API_2} from './src/ngtools_api';
```
Comment (tbosch) [inline packages/compiler-cli/ngtools.ts]: Could you actually copy the interfaces from `transformers/api.ts` into this file, and create wrapper functions for all functions in `entry_points`?
This way we get type errors should we change our interfaces when they don't match the interfaces for the cLI anymore / we are able to create adapters.

You can also remove all things from the interfaces in your copy of `transformers/api.ts` that the cli does not need.

Comment (tbosch) [reply packages/compiler-cli/ngtools.ts]: So probably create a new `src/ngtools_api2` with all the new interfaces and wrapper functions, and have a `compiler_cli/ngtools2.ts` from which you would import in the cli.

**Review by tbosch — CHANGES_REQUESTED** [review]

File: packages/compiler-cli/src/ngtools_api2.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+ * This is a private API for @ngtools/webpack. This API should be stable for NG 5.
+ *
+ * It contains copies of the interfaces needed and wrapper functions to ensure that use the copied
+ * interfaces to ensure they are not broken accidentally.
+ */
+
+import {ParseSourceSpan} from '@angular/compiler';
+import * as ts from 'typescript';
+
+export {NgTools_InternalApi_NG_2 as __NGTOOLS_PRIVATE_API_2} from './ngtools_api';
```
Comment (tbosch) [inline packages/compiler-cli/src/ngtools_api2.ts]: You don't need to do this export, as it is already reachable via `index`. I.e. we keep the old API for now as well.

File: packages/compiler-cli/ngtools2.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+export {CompilerHost, CustomTransformers, Diagnostic, EmitFlags, Program, __NGTOOLS_PRIVATE_API_2, createCompilerHost, createProgram, formatDiagnostics} from './src/ngtools_api2';
```
Comment (tbosch) [inline packages/compiler-cli/ngtools2.ts]: remove `__NGTOOLS_PRIVATE_API_2`

File: packages/compiler-cli/src/ngtools_api2.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+/**
+ * This is a private API for @ngtools/webpack. This API should be stable for NG 5.
+ *
+ * It contains copies of the interfaces needed and wrapper functions to ensure that use the copied
```
Comment (tbosch) [inline packages/compiler-cli/src/ngtools_api2.ts]: typo:  2x ensure

Comment (tbosch) [reply packages/compiler-cli/src/ngtools_api2.ts]: Maybe also add: once the ngc api is public and stable, this can be removed.

**Review by tbosch — CHANGES_REQUESTED** [review]

**Review by tbosch — APPROVED** [review]

File: packages/compiler-cli/src/ngtools_api2.ts [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+// Interfaces from ./transformers/api;
+export interface Diagnostic {
+  message: string;
+  span?: ParseSourceSpan;
+  category: ts.DiagnosticCategory;
+}
+
+export interface CompilerOptions extends ts.CompilerOptions {
+  genDir?: string;
```
Comment (chuckjaz) [inline packages/compiler-cli/src/ngtools_api2.ts]: `genDir` should be removed as it is not used starting with v5

File: packages/compiler-cli/src/ngtools_api2.ts [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+  emit({emitFlags, cancellationToken, customTransformers, emitCallback}: {
+    emitFlags?: EmitFlags,
+    cancellationToken?: ts.CancellationToken,
+    customTransformers?: CustomTransformers,
+    emitCallback?: TsEmitCallback
+  }): ts.EmitResult;
+}
+
+// Wrapper for createProgram.
+import {createProgram as createProgramOrig} from './transformers/program';
```
Comment (chuckjaz) [inline packages/compiler-cli/src/ngtools_api2.ts]: Move imports to the top of the module.

File: packages/compiler-cli/src/ngtools_api2.ts [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+export function createProgram(
+    {rootNames, options, host, oldProgram}:
+        {rootNames: string[], options: CompilerOptions, host: CompilerHost, oldProgram?: Program}):
+    Program {
+  return createProgramOrig({rootNames, options, host, oldProgram});
+}
+
+// Wrapper for createCompilerHost.
+import {createCompilerHost as createCompilerOrig} from './transformers/compiler_host';
```
Comment (chuckjaz) [inline packages/compiler-cli/src/ngtools_api2.ts]: Same here.

File: packages/compiler-cli/src/ngtools_api2.ts [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+export function createCompilerHost(
+    {options, tsHost = ts.createCompilerHost(options, true)}:
+        {options: CompilerOptions, tsHost?: ts.CompilerHost}): CompilerHost {
+  return createCompilerOrig({options, tsHost});
+}
+
+// Wrapper for formatDiagnostics.
+import {formatDiagnostics as formatDiagnosticsOrig} from './perform-compile';
```
Comment (chuckjaz) [inline packages/compiler-cli/src/ngtools_api2.ts]: and here.

**Comment by filipesilva** [comment]
@chuckjaz added the changes you requested.

**Review by jasonaden — APPROVED** [review]

**Pull request closed without merging**